Investigate new RWC failures

We haven't merged our RWC failures from the last 9 days and they've _really_ added up (which is, I assume, why nobody is looking at them or merging them). _Most_ of them are benign - error span/message changes for the new overload errors, some expected duplicate definition errors related to the `Generator` types change and older `node` declaration files. _However_, there's one pattern of changes alone these lines (mostly within s4l):
```
        protected _buildState(props: P, initialBuild: boolean): Partial<S> {
            const newState = {
                optionToPresent: this._getOptionFromConf()
            } as S;
            return newState;
            ~~~~~~~~~~~~~~~~
!!! error TS2322: Type 'S' is not assignable to type 'Partial<S>'.
!!! error TS2322:   Type 'ILayoutExperimentState' is not assignable to type 'Partial<S>'.
        }
```
that's suuuuuper concerning, since a `T` is pretty much always supposed to be assignable to a `Partial<T>`, meaning we've broken something pretty important ❤️ 

I've merged the current baselines so we can actually browse thru the diffs again, but this particular regression needs to be seriously investigated and addressed. Best guess is that #32071 interacted poorly with another change added to `master` (or the other failure in `master` hid this one) over the same period, but I dunno.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -13443,7 +13443,8 @@ namespace ts {
                             if (includeOptional
                                     ? !(filteredByApplicability!.flags & TypeFlags.Never)
                                     : isRelatedTo(targetConstraint, sourceKeys)) {
-                                const indexingType = filteredByApplicability || getTypeParameterFromMappedType(target);
+                                const typeParameter = getTypeParameterFromMappedType(target);
+                                const indexingType = filteredByApplicability ? getIntersectionType([filteredByApplicability, typeParameter]) : typeParameter;
                                 const indexedAccessType = getIndexedAccessType(source, indexingType);
                                 const templateType = getTemplateTypeFromMappedType(target);
                                 if (result = isRelatedTo(indexedAccessType, templateType, reportErrors)) {

--- a/tests/baselines/reference/mappedTypeRelationships.errors.txt
+++ b/tests/baselines/reference/mappedTypeRelationships.errors.txt
@@ -393,3 +393,7 @@ tests/cases/conformance/types/mapped/mappedTypeRelationships.ts(168,5): error TS
         const n: Numeric<T> = { x: 1 };
     }
     
+    function f<T extends { x: {} }>(): Partial<T> {
+        return undefined! as T;
+    }
+    

--- a/tests/baselines/reference/mappedTypeRelationships.js
+++ b/tests/baselines/reference/mappedTypeRelationships.js
@@ -187,6 +187,10 @@ function f90<T extends { x: number }>() {
     const n: Numeric<T> = { x: 1 };
 }
 
+function f<T extends { x: {} }>(): Partial<T> {
+    return undefined! as T;
+}
+
 
 //// [mappedTypeRelationships.js]
 function f1(x, k) {
@@ -319,6 +323,9 @@ function f82(t, k1, k2) {
 function f90() {
     var n = { x: 1 };
 }
+function f() {
+    return undefined;
+}
 
 
 //// [mappedTypeRelationships.d.ts]
@@ -408,3 +415,6 @@ declare type Numeric<T> = {
 declare function f90<T extends {
     x: number;
 }>(): void;
+declare function f<T extends {
+    x: {};
+}>(): Partial<T>;

--- a/tests/baselines/reference/mappedTypeRelationships.symbols
+++ b/tests/baselines/reference/mappedTypeRelationships.symbols
@@ -824,3 +824,15 @@ function f90<T extends { x: number }>() {
 >x : Symbol(x, Decl(mappedTypeRelationships.ts, 185, 27))
 }
 
+function f<T extends { x: {} }>(): Partial<T> {
+>f : Symbol(f, Decl(mappedTypeRelationships.ts, 186, 1))
+>T : Symbol(T, Decl(mappedTypeRelationships.ts, 188, 11))
+>x : Symbol(x, Decl(mappedTypeRelationships.ts, 188, 22))
+>Partial : Symbol(Partial, Decl(lib.es5.d.ts, --, --))
+>T : Symbol(T, Decl(mappedTypeRelationships.ts, 188, 11))
+
+    return undefined! as T;
+>undefined : Symbol(undefined)
+>T : Symbol(T, Decl(mappedTypeRelationships.ts, 188, 11))
+}
+

--- a/tests/baselines/reference/mappedTypeRelationships.types
+++ b/tests/baselines/reference/mappedTypeRelationships.types
@@ -668,3 +668,13 @@ function f90<T extends { x: number }>() {
 >1 : 1
 }
 
+function f<T extends { x: {} }>(): Partial<T> {
+>f : <T extends { x: {}; }>() => Partial<T>
+>x : {}
+
+    return undefined! as T;
+>undefined! as T : T
+>undefined! : never
+>undefined : undefined
+}
+

--- a/tests/cases/conformance/types/mapped/mappedTypeRelationships.ts
+++ b/tests/cases/conformance/types/mapped/mappedTypeRelationships.ts
@@ -188,3 +188,7 @@ type Numeric<T> = { [K in keyof T]?: number };
 function f90<T extends { x: number }>() {
     const n: Numeric<T> = { x: 1 };
 }
+
+function f<T extends { x: {} }>(): Partial<T> {
+    return undefined! as T;
+}

